File: leetcode/2571. Minimum Operations to Reduce an Integer to 0/2571.py
# https://helloacm.com/teaching-kids-programming-minimum-operations-to-reduce-an-integer-to-0-greedy-recursion-top-down-dynamic-programming-algorithm/
# https://leetcode.com/problems/minimum-operations-to-reduce-an-integer-to-0/
# MEDIUM, DP, RECURSION, MATH

class Solution:
    def minOperations(self, n: int) -> int:
        @cache
        def f(n):
            if n == 0:
                return 0
            if (n & (n - 1)) == 0:
                return 1
            a = int(log2(n))
            b = a + 1
            return f(min(n - (1 << a), (1 << b) - n)) + 1

        return f(n)

File: leetcode/2571. Minimum Operations to Reduce an Integer to 0/2571.rs
// https://helloacm.com/teaching-kids-programming-minimum-operations-to-reduce-an-integer-to-0-greedy-iterative-algorithm/
// https://leetcode.com/problems/minimum-operations-to-reduce-an-integer-to-0/
// MEDIUM, GREEDY, MATH

impl Solution {
    pub fn min_operations(mut n: i32) -> i32 {
        let mut ans = 0;
        while (n > 0) {
            let x = (n as f32).log2().floor() as i32;
            let na = n - (1 << x);
            let nb = (1 << (x + 1)) - n;
            n = na.min(nb);
            ans += 1;
        }
        ans        
    }
}
